os/arch/arm/src/amebad/amebad_watchdog_lowerhalf.c

```diff
@@ -0,0 +1,482 @@
+ *
+ * Input Parameters:
+ *   lower   - A pointer the publicly visible representation of the
+ *             "lower-half" driver state structure.
+ *   timeout - The new timeout value in milliseconds.
+ *
+ * Returned Values:
+ *   Zero on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+static int amebad_wdg_settimeout(FAR struct watchdog_lowerhalf_s *lower, uint32_t timeout)
+{
+	struct amebad_wdg_lowerhalf_s *priv = (struct amebad_wdg_lowerhalf_s *)lower;
+	DEBUGASSERT(priv);
+	if (!priv) {
+		return -ENODEV;
+	}
+
+	/*  Reset the watchdog timer with new timeout value */
+	priv->timeout_ms = timeout;
+	watchdog_init(priv->timeout_ms);
+	if (priv->int_mode) {
+		watchdog_irq_init((void *)priv->handler, (uint32_t)priv);
+	}
+
+	/* Re-start watchdog only if the status is WDT_START */
+	if (priv->wdt_status == WDT_START) {
+		watchdog_start();
+	}
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: amebad_wdg_capture
+ *
+ * Description:
+ *   Don't reset on watchdog timer timeout; instead, call this user provider
+ *   timeout handler.  NOTE:  Providing handler==NULL will restore the reset
+ *   behavior.
+ *
+ * Input Parameters:
+ *   lower      - A pointer the publicly visible representation of the "lower-half"
+ *                driver state structure.
+ *   newhandler - The new watchdog expiration function pointer.  If this
+ *                function pointer is NULL, then the reset-on-expiration
+ *                behavior is restored,
+ *
+ * Returned Values:
+ *   The previous watchdog expiration function pointer or NULL is there was
+ *   no previous function pointer, i.e., if the previous behavior was
+ *   reset-on-expiration (NULL is also returned if an error occurs).
+ *
+ ****************************************************************************/
+static xcpt_t amebad_wdg_capture(FAR struct watchdog_lowerhalf_s *lower, xcpt_t handler)
+{
+	struct amebad_wdg_lowerhalf_s *priv = (struct amebad_wdg_lowerhalf_s *)lower;
+	xcpt_t oldhandler;
+	DEBUGASSERT(priv);
+	if (!priv) {
+		return NULL;
+	}
+
+	/* Record previous handler as the oldhandler */
+	oldhandler = priv->handler;
+
+	/* Set the new handler */
+	priv->handler = handler;
+
+	/* Set to default RESET_MODE first */
+	if (priv->wdt_status == WDT_START) {	// Skip the timeout_left calculation if the status is ON_PAUSE
+		priv->timeout_left = priv->timeout_ms - ((SYSTIMER_TickGet() - priv->start_tick - priv->on_pause_duration) * 31) / 1000;
+	}
+	watchdog_init(priv->timeout_left);	// Re-init watchdog with remaining timeout
+	priv->int_mode = false;
+
+	/* Change to INT_MODE if handler exists */
+	if (handler) {
+		watchdog_irq_init((void *)priv->handler, (uint32_t)priv);	// INT_MODE enabled
+		priv->int_mode = true;
+	}
+
+	/* Re-start watchdog only if the status is WDT_START */
+	if (priv->wdt_status == WDT_START) {
+		watchdog_start();
```

Comment:
You mentioned `Re-start watchdog`. Is it okay to call `watchdog_start`, not `watchdog_refresh`?

Reply:
When using '**watchdog_init** then **watchdog_start**', the watchdog will be reset with new timeout value (here using timeout_left). It is the same function as refresh.